Fix possible hats typos, comment other proposals

badlop opened the pull request:
I'm implementing Hats 0.3.0 in ejabberd, and I've noticed some incoherences in the XEP. As I imagine they are not intended, this PR includes those changes to open discussion:

- Say "room" instead of "service" when the action deals with a specific room
- In Example 1, instead of Romeo, use the same User JID than the other examples
- In Example 1, sort attributes
- Fix capitalization of "Hat title" label
- In examples, use the defined field var "hats#uri" instead of ambiguous "hat"

---

Other proposals that I'm thinking about, but are not included in this PR:

### Destroy Hat

When destroying a hat, right now it asks the hat URI. However, as it is only possible to destroy a hat that already exists, the formulary offered by the server could instead provide the list of hats, then the client picks one of them to destroy.

This would be similar to the formulary to assign a hat to a user.

### Move 2.1

I would say that "2.1 Listing Hats" should be moved down to section "3. Protocol"

### List Assigned Hats

And finally, that XEP still does not provide any way to list the **assigned** hats, as described in [ejabberd custom XEP List Hats](https://docs.ejabberd.im/tutorials/muc-hats/#admin-requests-to-list-hats). Maybe that feature is not really needed in practice by clients? 

If it were needed, then the XEP could get another command: "List Assigned Hats", which returns a list of `{hats#jid, hats#uri}` that are assigned in the room

badlop committed:
Fix some typos

- Say "room" instead of "service" when the action deals with a specific room
- In Example 1, instead of Romeo, use the same User JID than the other examples
- In Example 1, sort attributes
- Fix capitalization of "Hat title" label
- In examples, use the defined field var "hats#uri" instead of ambiguous "hat"
- Fix duplicate "should" wordTODO:
- "2.1 Listing Hats" should be moved down to section "3. Protocol"
- Should there be a new command "List Assigned Hats",
  which returns a list of {hats#jid, hats#uri} that are assigned in the room

edhelas commented:
Hi,

I'd like to thank you personally for the awesome work you did on the ejabberd implementation... and this feedback PR that I wasn't expecting. 

Regarding the changes you've made in the PR those all look valid to me and I'm really happy that you're fixing them :)

## 3.3 Regarding Destroying a Hat

The section  only specify the set example and could indeed also show the example to request the form to destroy a Hat, that could then include the Hats list. If you want to can add those 2 examples to your PR:

```xml
<iq from='professor@example.edu/office'
    id='gd53a2b6'
    to='physicsforpoets@courses.example.edu'
    type='set'
    xml:lang='en'>
  <command xmlns='http://jabber.org/protocol/commands'
            action='execute'
            node='urn:xmpp:hats:commands:destroy'/>
</iq>
```

```xml
<iq from='physicsforpoets@courses.example.edu'
    id='gd53a2b6'
    to='professor@example.edu/office'
    type='result'
    xml:lang='en'>
  <command xmlns='http://jabber.org/protocol/commands'
            node='urn:xmpp:hats:commands:destroy'
            sessionid='A971D19A-2226-4DAD-B261-9D0886B9A026'
            status='executing'>
    <x xmlns='jabber:x:data' type='form'>
      <title>Destroying a Hat</title>
      <instructions>Fill out this form to destroy a hat.</instructions>
      <field type='hidden' var='FORM_TYPE'>
        <value>urn:xmpp:hats:commands</value>
      </field>
      <field label='The role'
             type='list-single'
             var='hat'>
        <option label='Teacher'><value>http://tech.example.edu/hats#Teacher</value></option>
        <option label='Teacher&apos;s Assistant'><value>http://tech.example.edu/hats#TeacherAssistant</value></option>
        <option label='Test Proctor'><value>http://tech.example.edu/hats#Proctor</value></option>
      </field>
    </x>
  </command>
</iq>
```

And then the destruction itself:

```xml
<iq from='professor@example.edu/office'
    id='rei4n2b0'
    to='physicsforpoets@courses.example.edu'
    type='set'
    xml:lang='en'>
  <command xmlns='http://jabber.org/protocol/commands'
           node='urn:xmpp:hats:commands:destroy'>
    <x xmlns='jabber:x:data' type='submit'>
      <field type='hidden' var='FORM_TYPE'>
        <value>urn:xmpp:hats:commands</value>
      </field>
      <field var='hat'>
        <value>http://tech.example.edu/hats#TeacherAssistant</value>
      </field>
    </x>
  </command>
</iq>
```

What I'd like to be sure is that if its still possible to directly jump to the destruction without getting the form ? 

## Move 2.1

Go for it :) Its purely redaction but having the complete CRUD in one section looks good to me.

## List Assigned Hats

Indeed I didn't provide it as you can already get all the assigned hats from the room presence but if you feel its something that could be useful in some cases I'm not against adding a section for it.

I'll merge your PR once we are ok with everything, then ask the XSF to update the XEP, as it is purely redaction changes it shoudn't be a big issue.

Thanks again :) !

badlop committed:
Apply some changes commented in the PR

- Destroy Hat: Show example how to query form with list of hats
- Listing Hats: Move to the Protocol section
- Listing Assigned Hats: New section

badlop commented:
> If you want to can add those 2 examples to your PR:

Ok, examples added.

> What I'd like to be sure is that if its still possible to directly jump to the destruction without getting the form ?

Ok, I've added a paragraph mentioning that possibility.

> Move 2.1 Go for it :) Its purely redaction but having the complete CRUD in one section looks good to me.

Right: create, destroy, update, and listing are related operations.

> you can already get all the assigned hats from the room presence

Disclaimer: I only comment this thinking in potential users, but as I don't use that XEP myself, maybe that topic is not so important. I am just speculative and conceptually thinking:

You don't get **_all_** the assigned hats from room presences: you only get the assigned hats **_for users that send presence stanzas_**.

The occupant client gets the assigned hats in presence... when those users change presence.

And the admin client does too. But the admin client does not get all the assigned hats, only the ones that users change presence. How can an admin notice that a hat was wrongly assigned, or no longer needed, if that user never sends a presence stanza?

edhelas commented:
Thanks for the changes, I'll review them asap :)

I agree with you with your listing-all-assigned-hats proposal, if you want to add a new section with that one I'd be happy to have it, then we're really feature complete in the XEP.

badlop commented:
> I agree with you with your listing-all-assigned-hats proposal, if you want to add a new section with that one

It's already added in the last commit

edhelas commented:
<img width="720" height="720" alt="image" src="https://github.com/user-attachments/assets/73ecde82-23b5-4fb5-9ff6-63002c21ee2b" />

edhelas commented:
Here we go, I submitted you changes to the XSF there https://github.com/xsf/xeps/pull/1462 :) ! @badlop
